tensorflow/lite/experimental/kernels/unidirectional_sequence_gru.cc

```diff
@@ -0,0 +1,235 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/c/c_api_internal.h"
+#include "tensorflow/lite/experimental/kernels/gru_cell.h"
+#include "tensorflow/lite/kernels/internal/tensor.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+
+namespace tflite {
+namespace ops {
+namespace experimental {
+namespace unidirectional_sequence_gru {
+namespace {
+
+void GruImpl(const TfLiteTensor* input, const TfLiteTensor* input_state,
+             const TfLiteTensor* gate_weight, const TfLiteTensor* gate_bias,
+             const TfLiteTensor* candidate_weight,
+             const TfLiteTensor* candidate_bias, TfLiteTensor* output,
+             TfLiteTensor* output_state, TfLiteTensor* activation,
+             TfLiteTensor* concat) {
+  const int n_time = input->dims->data[0];
+  const int n_batch = input->dims->data[1];
+  const int n_input = input->dims->data[2];
+  const int n_output = output->dims->data[2];
+  const RuntimeShape input_shape = RuntimeShape({n_batch, n_input});
+  const float* input_data = GetTensorData<float>(input);
+  const RuntimeShape state_shape = GetTensorShape(input_state);
+  const float* input_state_data = GetTensorData<float>(input_state);
+  const RuntimeShape gate_weight_shape = GetTensorShape(gate_weight);
+  const float* gate_weight_data = GetTensorData<float>(gate_weight);
+  const RuntimeShape gate_bias_shape = GetTensorShape(gate_bias);
+  const float* gate_bias_data = GetTensorData<float>(gate_bias);
+  const RuntimeShape candidate_weight_shape = GetTensorShape(candidate_weight);
+  const float* candidate_weight_data = GetTensorData<float>(candidate_weight);
+  const RuntimeShape candidate_bias_shape = GetTensorShape(candidate_bias);
+  const float* candidate_bias_data = GetTensorData<float>(candidate_bias);
+  const RuntimeShape activation_shape = GetTensorShape(activation);
+  const RuntimeShape output_shape = RuntimeShape({n_batch, n_output});
+  float* output_data = GetTensorData<float>(output);
+  float* output_state_data = GetTensorData<float>(output_state);
+  float* activation_data = GetTensorData<float>(activation);
+  const RuntimeShape concat_shape = GetTensorShape(concat);
+  float* concat_data = GetTensorData<float>(concat);
+  for (int i = 0; i < n_time; ++i) {
+    gru_cell::GruCell(
+        input_shape, input_data, state_shape, input_state_data,
+        gate_weight_shape, gate_weight_data, gate_bias_shape, gate_bias_data,
+        candidate_weight_shape, candidate_weight_data, candidate_bias_shape,
+        candidate_bias_data, output_shape, output_data, output_state_data,
+        activation_shape, activation_data, concat_shape, concat_data);
+    input_data += n_batch * n_input;
```

Comment:
another nit:

n_batch * n_input & n_batch * n_output can be computed outside the loop, maybe call it, input_offset & output_offset

Reply:
Done